File: statistics/src/stats.py
# To work with dates
from datetime import datetime, timedelta

from src.database import getDatabase


def getStatistics():
    data = getDatabase('users', 'user-data', log=False)['users']
    accountCreations = [returnDate(value) for value in data.values()]
    accountCreations = list(filter(lambda value: len(value) != 0, accountCreations))
    creationStatistics = returnCreationCount(accountCreations)

    return creationStatistics


def returnCreationCount(accountCreations):
    earliestDate = datetime.strptime(accountCreations[0], "%Y-%m-%d")
    latestDate = datetime.strptime(accountCreations[-1], "%Y-%m-%d")
    dates = getDatesInRange(earliestDate, latestDate)

    creationStatistics = []

    for date in accountCreations:
        if len(creationStatistics) != 0 and date == creationStatistics[-1][0]:
            creationStatistics[-1][1] += 1
        else:
            creationStatistics.append([date, 1])

    for i in range(len(dates)):
        if dates[i] != creationStatistics[i][0]:
            creationStatistics.insert(i, [dates[i], 0])

    return creationStatistics


def getDatesInRange(startDate, endDate):
    delta = timedelta(days=1)
    dates = []

    while startDate <= endDate:
        dates.append(startDate.isoformat()[0:10])
        startDate += delta

    return dates


def returnDate(value):
    date = value.get('created', '')

    if date == '' or len(date) == 0:
        return ''
    else:
        return date[0:10]

File: statistics/stats.js
async function getStatistics() {
    const data = await getDatabase();
    const accountCreations = returnDates(data);
    const creationStatistics = returnCreationCount(accountCreations);

    return creationStatistics
}

function returnCreationCount(accountCreations) {
    const earliestDate = new Date(accountCreations[0]);
    const latestDate = new Date();
    const dates = getDatesInRange(earliestDate, latestDate);

    let creationStatistics = [];

    for (let i = 0; i < accountCreations.length; i++) {
        let date = accountCreations[i];

        if (creationStatistics.length != 0 && date == creationStatistics[creationStatistics.length-1][0]) {
            creationStatistics[creationStatistics.length-1][1] += 1
        } else {
            creationStatistics.push([date, 1])
        }
    }

    for (let i = 0; i < dates.length; i++) {
        let date;

        try {
            date = creationStatistics[i][0]
        } catch {
            creationStatistics.splice(i, 0, [dates[i], 0])
            continue
        }

        if (dates[i] != date) {
            creationStatistics.splice(i, 0, [dates[i], 0])
        }
    }

    return creationStatistics
}


Date.prototype.addDays = function(days) {
    let date = new Date(this.valueOf());
    date.setDate(date.getDate() + days);
    return date;
}

function getDatesInRange(startDate, stopDate) {
    let dates = [];
    let currentDate = startDate;
    while (currentDate <= stopDate) {
        dates.push(new Date (currentDate).toISOString().substring(0, 10));
        currentDate = currentDate.addDays(1);
    }
    return dates;
}


function returnDates(data) {
    let dates = [];

    for (let key in data) {
        let date;

        date = data[key].created;

        if (date == undefined || date.length == 0) {
            continue
        } else {
            dates.push(date.substring(0, 10))
        }
    }

    return dates
}
